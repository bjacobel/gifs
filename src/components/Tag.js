import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import classNames from 'classnames';

import {
  addTagAsync,
  deleteTagAsync,
} from '../actions/tags';
import { isAuthed } from '../services/auth0';

const mapStateToProps = state => ({
  auth: state.auth,
});

const mapDispatchToProps = {
  addTagAsync,
  deleteTagAsync,
};

class Tag extends Component {
  componentWillMount() {
    this.setState({ value: '' });
  }

  render() {
    const {
      content,
      id,
      meta,
      /* eslint-disable no-shadow */
      addTagAsync,
      deleteTagAsync,
      /* eslint-enable no-shadow */
      auth,
    } = this.props;

    const updateTagToAdd = event => this.setState({ value: event.target.value });
    const deleteTag = () => deleteTagAsync(id);
    const addNewTag = () => {
      if (this.state.value) {
        addTagAsync(this.state.value);
        this.setState({ value: '' });
      }
    };
    const submitOnEnter = (event) => {
      if (event.nativeEvent.keyCode === 13) {
        addNewTag();
      }
    };

    const tagInput = (
      <input
        autoCapitalize="off"
        autoComplete="off"
        autoCorrect="off"
        autoFocus
        className="tag-to-add"
        onChange={ updateTagToAdd }
        onKeyPress={ submitOnEnter }
        placeholder="Add tag"
        spellCheck="false"
        type="text"
        value={ this.state.value }
        disabled={ !isAuthed(auth) }
      />
    );

    let addOrDel;

    if (meta === 'add-tag') {
      addOrDel = (
        <span>
          { tagInput }
          <span className="add-tag" onMouseUp={ addNewTag }><b>＋</b></span>
        </span>
      );
    } else {
      addOrDel = <span className="del-tag" onMouseUp={ deleteTag }>✕</span>;
    }

    return (
      <span className={ classNames('tag-wrapper', { disabled: !isAuthed(auth), adder: meta === 'add-tag' }) }>
        <span className="tag">
          { content }
          { addOrDel }
        </span>
        { /* eslint-disable react/self-closing-comp */ }
        <span className="break"> </span>
        { /* eslint-enable react/self-closing-comp */ }
      </span>
    );
  }
}

Tag.propTypes = {
  content: PropTypes.string,
  id: PropTypes.string,
  meta: PropTypes.oneOf(['add-tag', undefined]),
  auth: PropTypes.shape({
    idToken: PropTypes.string,
    idTokenExpiry: PropTypes.number,
  }),
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Tag);
