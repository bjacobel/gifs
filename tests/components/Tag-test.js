import React from 'react';
import { shallow } from 'enzyme';

import Tag from '../../src/components/Tag';

describe('Tag component', () => {
  describe('when unauthenticated', () => {
    it('matches snapshot if displayed as an add-tag', () => {
      expect(shallow(
        <Tag.WrappedComponent
          meta="add-tag"
          auth={ { isAuthenticated: false } }
          addTagAsync={ jest.fn() }
          deleteTagAsync={ jest.fn() }
        />,
      )).toMatchSnapshot();
    });

    it('matches snapshot if displayed as a normal tag', () => {
      expect(shallow(
        <Tag.WrappedComponent
          content="first tag"
          id="1"
          auth={ { isAuthenticated: false } }
          addTagAsync={ jest.fn() }
          deleteTagAsync={ jest.fn() }
        />,
      )).toMatchSnapshot();
    });
  });

  describe('when authenticated', () => {
    describe('a normal tag', () => {
      it('matches snapshot', () => {
        expect(shallow(
          <Tag.WrappedComponent
            content="first tag"
            id="1"
            auth={ { isAuthenticated: true } }
            addTagAsync={ jest.fn() }
            deleteTagAsync={ jest.fn() }
          />,
        )).toMatchSnapshot();
      });

      it('fires delete tag method when del-tag span clicked', () => {
        const deleteTagAsync = jest.fn();
        const enzymeRepr = shallow(
          <Tag.WrappedComponent
            content="first tag"
            id="1"
            auth={ { isAuthenticated: true } }
            addTagAsync={ jest.fn() }
            deleteTagAsync={ deleteTagAsync }
          />,
        );

        enzymeRepr.find('.del-tag').simulate('mouseUp');
        expect(deleteTagAsync).toBeCalledWith('1');
      });
    });

    describe('an add-tag', () => {
      it('matches snapshot', () => {
        expect(shallow(
          <Tag.WrappedComponent
            meta="add-tag"
            auth={ { isAuthenticated: true } }
            addTagAsync={ jest.fn() }
            deleteTagAsync={ jest.fn() }
          />,
        )).toMatchSnapshot();
      });

      it('updates state with the content to add on change', () => {
        const enzymeRepr = shallow(
          <Tag.WrappedComponent
            meta="add-tag"
            auth={ { isAuthenticated: true } }
            addTagAsync={ jest.fn() }
            deleteTagAsync={ jest.fn() }
          />,
        );

        enzymeRepr.find('input').simulate('change', { target: { value: 'tag to add' } });
        expect(enzymeRepr.state('value')).toEqual('tag to add');
      });

      it('adds the tag when the add-tag span is clicked', () => {
        const addTagAsync = jest.fn();
        const enzymeRepr = shallow(
          <Tag.WrappedComponent
            meta="add-tag"
            auth={ { isAuthenticated: true } }
            addTagAsync={ addTagAsync }
            deleteTagAsync={ jest.fn() }
          />,
        );

        enzymeRepr.setState({ value: 'tag to add' });
        enzymeRepr.find('.add-tag').simulate('mouseUp');
        expect(addTagAsync).toBeCalledWith('tag to add');
      });

      it('adds the tag when enter is pressed', () => {
        const addTagAsync = jest.fn();
        const enzymeRepr = shallow(
          <Tag.WrappedComponent
            meta="add-tag"
            auth={ { isAuthenticated: true } }
            addTagAsync={ addTagAsync }
            deleteTagAsync={ jest.fn() }
          />,
        );

        enzymeRepr.setState({ value: 'tag to add' });
        enzymeRepr.find('input').simulate('keyPress', { nativeEvent: { keyCode: 13 } });
        expect(addTagAsync).toBeCalledWith('tag to add');
      });
    });
  });
});
