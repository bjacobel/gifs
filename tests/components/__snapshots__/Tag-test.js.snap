exports[`Tag component when authenticated a normal tag matches snapshot 1`] = `
<span
  className="tag-wrapper">
  <span
    className="tag">
    first tag
    <span
      className="del-tag"
      onMouseUp={[Function]}>
      ✕
    </span>
  </span>
  <span
    className="break">
     
  </span>
</span>
`;

exports[`Tag component when authenticated an add-tag matches snapshot 1`] = `
<span
  className="tag-wrapper adder">
  <span
    className="tag">
    <span>
      <input
        autoCapitalize="off"
        autoComplete="off"
        autoCorrect="off"
        autoFocus={true}
        className="tag-to-add"
        disabled={false}
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder="Add tag"
        spellCheck="false"
        type="text"
        value="" />
      <span
        className="add-tag"
        onMouseUp={[Function]}>
        <b>
          ＋
        </b>
      </span>
    </span>
  </span>
  <span
    className="break">
     
  </span>
</span>
`;

exports[`Tag component when unauthenticated matches snapshot if displayed as a normal tag 1`] = `
<span
  className="tag-wrapper disabled">
  <span
    className="tag">
    first tag
    <span
      className="del-tag"
      onMouseUp={[Function]}>
      ✕
    </span>
  </span>
  <span
    className="break">
     
  </span>
</span>
`;

exports[`Tag component when unauthenticated matches snapshot if displayed as an add-tag 1`] = `
<span
  className="tag-wrapper disabled adder">
  <span
    className="tag">
    <span>
      <input
        autoCapitalize="off"
        autoComplete="off"
        autoCorrect="off"
        autoFocus={true}
        className="tag-to-add"
        disabled={true}
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder="Add tag"
        spellCheck="false"
        type="text"
        value="" />
      <span
        className="add-tag"
        onMouseUp={[Function]}>
        <b>
          ＋
        </b>
      </span>
    </span>
  </span>
  <span
    className="break">
     
  </span>
</span>
`;
